src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -2162,6 +2162,288 @@ private int codePointX(Encoding enc, CodeRange codeRange, byte[] bytes, int p, i
         }
     }
 
+    @CoreMethod(names = "undump", taintFrom = 0)
+    @ImportStatic(StringGuards.class)
+    public abstract static class UndumpNode extends CoreMethodArrayArgumentsNode {
+        @Specialization(guards = "isAsciiCompatible(string)")
+        protected RubyString undumpAsciiCompatible(RubyString string,
+                @CachedLanguage RubyLanguage language,
+                @Cached AllocateHelperNode allocateHelperNode,
+                @Cached RopeNodes.MakeLeafRopeNode makeLeafRopeNode) {
+            // Taken from org.jruby.RubyString#undump
+            RopeBuilder outputBytes = undump(string.rope);
+
+            final Rope rope = makeLeafRopeNode
+                    .executeMake(
+                            outputBytes.getBytes(),
+                            outputBytes.getEncoding(),
+                            CR_UNKNOWN,
+                            outputBytes.getLength());
+
+            final RubyClass logicalClass = coreLibrary().stringClass;
+            final Shape shape = allocateHelperNode.getCachedShape(logicalClass);
+            final RubyString result = new RubyString(logicalClass, shape, false, false, rope);
+            allocateHelperNode.trace(result, this, language);
+            return result;
+        }
+
+        @Specialization(guards = "!isAsciiCompatible(string)")
+        protected RubyString undumpNonAsciiCompatible(RubyString string) {
+            throw new RaiseException(
+                    getContext(),
+                    getContext().getCoreExceptions().encodingCompatibilityError(
+                            "ASCII incompatible encoding: " + string.rope.encoding,
+                            this));
+        }
+
+        private RaiseException newRuntimeError(String message) {
+            return new RaiseException(getContext(), getContext().getCoreExceptions().runtimeError(message, this));
+        }
+
+        private RaiseException invalidFormat() {
+            return newRuntimeError(
+                    "invalid dumped string; not wrapped with '\"' nor '\"...\".force_encoding(\"...\")' form");
+        }
+
+        private static final byte[] FORCE_ENCODING_BYTES = ".force_encoding(\"".getBytes();
+
+        @TruffleBoundary
+        private RopeBuilder undump(Rope rope) {
+            byte[] bytes = rope.getBytes();
+            int[] start = { 0 };
```

Comment:
Could you use a simple `int` for `start`?
I think if `undumpAfterBackslash()` returns the new start, then it should work.

Reply:
Yeah, that should work, will fix.